### PR TITLE
Rank a word-boundary after a non-word-boundary

### PR DESCRIPTION
**Pull request by ganwell**
Update: removed ideas that turned out to be wrong.

### CODE REVIEW

**Comment by ganwell** [comment]
Refs #3015

File: src/ranked_match.cc [hunk 1]
```diff
@@ -36,9 +36,11 @@ bool matches(UsedLetters query, UsedLetters letters)
 
 using Utf8It = utf8::iterator<const char*>;
 
-static int count_word_boundaries_match(StringView candidate, StringView query)
+static int count_word_boundaries_match_first(StringView candidate, StringView query, int &first_word_len)
```
Comment (mawww) [inline src/ranked_match.cc]: declare a struct WordBoundaries { int count; int first_word_len; } and return that, instead of an out parameter.

File: src/ranked_match.cc [hunk 2]
```diff
@@ -68,6 +75,12 @@ static int count_word_boundaries_match(StringView candidate, StringView query)
     }
     return count;
 }
+static int count_word_boundaries_match(StringView candidate, StringView query)
```
Comment (mawww) [inline src/ranked_match.cc]: No need for this function.

File: src/ranked_match.cc [hunk 3]
```diff
@@ -228,6 +238,9 @@ bool RankedMatch::operator<(const RankedMatch& other) const
 }
 
 UnitTest test_ranked_match{[] {
+    int first_word_len = 0;
+    count_word_boundaries_match_first("run(test)", "rat", first_word_len);
```
Comment (mawww) [inline src/ranked_match.cc]:
```
auto eq = [](WordBoundaries lhs, WordBoundaries rhs) { return lhs.count == rhs.count and lhs.first_word_len == rhs.first_word_len; };
kak_assert(eq(count_word_boundaries_match(...), {2, 1}))
...
```

**Comment by ganwell** [comment]
Thanks a lot for the tips, I really like what C++ looks like in the kakoune-source.

I first thought that Flags::SingleWord is related to the whole candidate, not only the match. So I had to count word-boundaries for whole candidate, too. If we don't check for the boundary-count, the preference for short words is too strong.

```cpp
    if (first_word_len != other_first_word_len)
        return first_word_len < other_first_word_len;
```

```
hal
Hal                     *s:en_GB*█
hale                    *s:en_GB*░
half                    *s:en_GB*░
hall                    *s:en_GB*░
halo                    *s:en_GB*░
halt                    *s:en_GB*░
Hale                    *s:en_GB*░
Hall                    *s:en_GB*░
Hals                    *s:en_GB*░
halal                   *s:en_GB*░
```

If we only use the first word length, if we match a multi-word, the single-word case stays the same (of course).

```cpp
    if ((count > 1 or other_count > 1) and first_word_len != other_first_word_len)
        return first_word_len < other_first_word_len;
```

```
hal
Hal                     *s:en_GB*█
halal                   *s:en_GB*░
halberd                 *s:en_GB*░
halcyon                 *s:en_GB*░
hale                    *s:en_GB*░
half                    *s:en_GB*░
halfback                *s:en_GB*░
halfhearted             *s:en_GB*░
halfheartedness         *s:en_GB*░
halfpence               *s:en_GB*░
```

I am starting to understand the RankedMatch better, but I still don't get the `find(1p)`/`findfs(8)` use-case completely. Especially because one can complete camelCase and snake_case, but not `find(1p)` the candidate is truncated at `(`. I more or less worked till the tests stopped failing.

So I like the code much better than before, but I am still not sure about the logic.

**Comment by ganwell** [comment]
TLDR: I will write a better solution, using actual boundaries instead of word length.

I discussed the PR with a friend and I now understand the use-case and I know why the current logic isn't right.

It works for a basic case:

```
3│find(1p)
2│findfs(8)
1│
4│fin
~ find(1p)  *scratch*█
~ findfs(8) *scratch*█
```

But not for a more general case:

```
3│bla_find(1p)
2│bla_findfs(8)
1│bla_
4│bla_fin
~ bla_findfs(8) *scratch*█
~ bla_find(1p)  *scratch*█
```

I think, I also found a pretty straight forward way to implement this in a much smaller patch. I replace

```if (punct1 != punct2)```

with

```if (is_boundary1 != is_boundary2)```

Where I have to write code to set is_boundary by using the last codepoint too.

**Comment by mawww** [comment]
I had a WIP implementation for this that implemented the new behaviour you propose, it seemed to be working nicely at the time, I can either get it back, or let you implement that. It is a bit trickier than it appears due to utf8 decoding, as we first look at byte mismatches then switch to decoded codepoint at the mismatch position.

**Comment by ganwell** [comment]
@mawww you say your implementation is WIP? If its easy to finish, I'd just push it. It would be interesting to add the "bla_find(1)"-case to the asserts, though. If there is still work to do, I could try to finish it, just point to the brach/commit.

**Comment by ganwell** [comment]
@mawww I updated the PR, see the description on top https://github.com/mawww/kakoune/pull/3038#issue-302156615

File: src/ranked_match.cc [hunk 4]
```diff
@@ -207,14 +223,19 @@ bool RankedMatch::operator<(const RankedMatch& other) const
...
         if (cp1 != cp2)
         {
-            const bool punct1 = iswpunct((wchar_t)cp1);
-            const bool punct2 = iswpunct((wchar_t)cp2);
-            if (punct1 != punct2)
-                return punct1;
+            auto itlast1 = character_start_safe(itsave1 - 1, begin1);
+            auto itlast2 = character_start_safe(itsave2 - 1, begin2);
+            const auto cplast1 = utf8::read_codepoint(itlast1, end1);
+            const auto cplast2 = utf8::read_codepoint(itlast2, end2);
```
Comment (mawww) [inline src/ranked_match.cc]: I think this will never consider the first char as a word boundary, as character_safe_start will trigger comparison of this character against itself. I think a better alternative would be to make `cplastX = prev_codepoint(itlastX, beginX, endX).value_or(0)` with prev_codepoint returning an Optional<Codepoint> when it exists.

Comment (ganwell) [reply src/ranked_match.cc]: * I implemented prev_codepoint and put it into the utf8 namespace, since it seems reusable

* I added cp_zero so my code doesn't add a warning or need a `(Codepoint) 0` cast. But let me know what you prefer.

* I added

```cpp
kak_assert(RankedMatch{"b(fa", "a"} < RankedMatch{"(bfa", "a"} );
```

Because one might think, following the logic of `find(1p)` this should be sorted the opposite way, but it is essentially the same as:

```cpp
kak_assert(RankedMatch{"init", ""} < RankedMatch{"__init__", ""});
```

Which is my initial case (#3015) and it is important for completing something like `sys.` in python. At least after trying for a while I didn't find a useful criteria to distinguish these cases.

* And of course I was wrong about combining the two is_word_boundary, I stashed away `debug =? yes`  The tests would fail and its also clear that the count needs directed word-boundaries.

File: src/utf8.hh [hunk 5]
```diff
@@ -258,6 +259,18 @@ Iterator character_start(Iterator it, const Sentinel& begin) noexcept
...
 
+// returns an optional iterator to the first byte of the previous character
+// or no value if it is at begin
+template<typename Iterator, typename Sentinel>
+static Optional<Codepoint> prev_codepoint(Iterator it, const Sentinel& begin) noexcept
+{
+    if (it <= begin)
+        return {};
+    auto itlast = utf8::character_start(it - 1, begin);
+    return utf8::read_codepoint(itlast, it);
```
Comment (mawww) [inline src/utf8.hh]: no need for utf8:: here as we already are in that namespace. Also, you can just do

```
return codepoint(character_start(it -1, begin), it);
```

Comment (ganwell) [reply src/utf8.hh]: @mawww Thank you for the patience, I should have seen these things.

File: src/unicode.hh [hunk 6]
```diff
@@ -12,6 +12,7 @@ namespace Kakoune
 {
 
 using Codepoint = char32_t;
+const static Codepoint cp_zero = 0;
```
Comment (mawww) [inline src/unicode.hh]: using `0` or `Codepoint{0}` would be as good. I'd rather not add magic constants like that.

Comment (ganwell) [reply src/unicode.hh]: I removed the constant. If there is any issue left - also formatting or style - let me know. I am not up to date with C++, it was 2005 when I last used it full-time.

Comment (ganwell) [reply src/unicode.hh]: Clang on travis threw an error because of `0`, I switched to `Codepoint{0}`

**Comment by mawww** [comment]
Thanks for working on that !
